contracts/paymaster/BasePaymaster.sol

```diff
@@ -0,0 +1,167 @@
+// SPDX-License-Identifier: GPL-3.0-only
+// COPIED FROM: https://github.com/opengsn/gsn/blob/master/packages/contracts/src/BasePaymaster.sol
+// Changes:
+// - Drops Ownable (OpenZeppelin).
+pragma solidity 0.8.10;
+
+import "@openzeppelin/contracts/utils/introspection/ERC165.sol";
+import "@openzeppelin/contracts/utils/introspection/ERC165Checker.sol";
+
+import "@opengsn/contracts/src/interfaces/IPaymaster.sol";
+import "@opengsn/contracts/src/interfaces/IRelayHub.sol";
+import "@opengsn/contracts/src/forwarder/IForwarder.sol";
+import "@opengsn/contracts/src/utils/GsnEip712Library.sol";
+
+/**
+ * @notice An abstract base class to be inherited by a concrete Paymaster.
+ * A subclass must implement:
+ *  - preRelayedCall
+ *  - postRelayedCall
+ */
+abstract contract BasePaymaster is IPaymaster, ERC165 {
+  using ERC165Checker for address;
+
+  IRelayHub internal relayHub;
+  address private _trustedForwarder;
+
+  /// @inheritdoc IPaymaster
+  function getRelayHub() public view override returns (address) {
+    return address(relayHub);
+  }
+
+  //overhead of forwarder verify+signature, plus hub overhead.
+  uint256 public constant FORWARDER_HUB_OVERHEAD = 50000;
+
+  //These parameters are documented in IPaymaster.GasAndDataLimits
+  uint256 public constant PRE_RELAYED_CALL_GAS_LIMIT = 100000;
+  uint256 public constant POST_RELAYED_CALL_GAS_LIMIT = 110000;
+  uint256 public constant PAYMASTER_ACCEPTANCE_BUDGET = PRE_RELAYED_CALL_GAS_LIMIT + FORWARDER_HUB_OVERHEAD;
+  uint256 public constant CALLDATA_SIZE_LIMIT = 10500;
+
+  /// @inheritdoc IERC165
+  function supportsInterface(bytes4 interfaceId) public view virtual override(IERC165, ERC165) returns (bool) {
+    return interfaceId == type(IPaymaster).interfaceId || super.supportsInterface(interfaceId);
+  }
+
+  /// @inheritdoc IPaymaster
+  function getGasAndDataLimits() public view virtual override returns (IPaymaster.GasAndDataLimits memory limits) {
+    return
+      IPaymaster.GasAndDataLimits(
+        PAYMASTER_ACCEPTANCE_BUDGET,
+        PRE_RELAYED_CALL_GAS_LIMIT,
+        POST_RELAYED_CALL_GAS_LIMIT,
+        CALLDATA_SIZE_LIMIT
+      );
+  }
+
+  /**
+   * @notice this method must be called from preRelayedCall to validate that the forwarder
+   * is approved by the paymaster as well as by the recipient contract.
+   */
+  function _verifyForwarder(GsnTypes.RelayRequest calldata relayRequest) internal view virtual {
+    require(getTrustedForwarder() == relayRequest.relayData.forwarder, "Forwarder is not trusted");
+    GsnEip712Library.verifyForwarderTrusted(relayRequest);
+  }
+
+  function _verifyRelayHubOnly() internal view virtual {
+    require(msg.sender == getRelayHub(), "can only be called by RelayHub");
```

Comment:
I would suggest that all these `require` are changed to `if` statements and proper `revert` exception primitives.
They give you more flexibility for tests (you can have exception take known arguments, and pass these as informations to the client or the test).